Use caching for `--python-downloads-json-url`

MeitarR opened the pull request:
Use the `CachedClient` for `--python-downloads-json-url` too, with the usual rules around freshness.

https://github.com/astral-sh/uv/pull/16542#discussion_r2522906275

MeitarR commented:
As commented at e4663d010e5997801c509f96757d91d84032201e I decided to bypass cache on `uv python list` and `uv python install` as it seems reasonable to me, but I leave the decision up to you whether to change it or always use the cache.

The reason for that is that I'm afraid that the file will change at the remote, and we will fetch the old file from the cache

konstin commented on crates/uv-python/src/downloads.rs:
```diff
@@ -1101,14 +1124,99 @@ impl ManagedPythonDownloadList {
 }
 
+/// Fetch Python downloads JSON metadata, using the cache when available.
+async fn read_cached_url(
```
Why does this use a custom implementation instead of using the existing caching utils? It seems we aren't actually using `CachedClient` here.

MeitarR replied:
You are right, I did not understand the CachedClient. I rewrote the PR after more research

konstin commented on crates/uv/tests/it/network.rs:
```diff
@@ -1101,3 +1101,59 @@ async fn retry_read_timeout_stream() {
+
+/// Check the python list error message when the server returns HTTP status 500, a retryable error.
+#[tokio::test]
+async fn python_list_remote_python_downloads_json_url_http_500() {
```
This fetch should use the existing utils and make this test redundant.

MeitarR committed:
Cache the remote Python downloads JSON via `CachedClient`

Switch the downloads JSON fetch from BaseClient to CachedClient so that
uv commands that only need the list for a prerelease warning or a
dry-run can hit the cache instead of going to the network every time.

konstin commented on crates/uv-python/src/downloads.rs:
```diff
@@ -1097,12 +1103,40 @@ impl ManagedPythonDownloadList {
+        "downloads-json",
+        format!("{}.msgpack", cache_digest(&url.as_str())),
+    );
+    let cache_control = CacheControl::from(cache.freshness(&cache_entry, None, None)?);
```
We should consider connectivity here and `AllowStale` in offline mode, similar to other callers of `freshness`

MeitarR replied:
good catch! fixed

MeitarR committed:
always AllowStale when offline

konstin committed:
Remove redundant comment

konstin committed:
Don't discard retry count

konstin committed:
Cache JSON instead of bytes

This avoids caching invalid JSON, and avoid the double deserialization cost from the cache.

konstin committed:
Only construct the client if we need it

We usually don't need this client, so we only construct it if it's requested.

konstin committed:
Simplify with match

konstin committed:
Improve error handling
